### PR TITLE
Allow data output streams to be generated via custom mechanisms when given partitioning and file name

### PR DESCRIPTION
**Pull request by mccheah**

### CODE REVIEW

File: api/src/main/java/com/netflix/iceberg/io/FileIO.java [hunk 1]
```diff
@@ -47,4 +46,30 @@
...
+   */
+  OutputFile newMetadataFile(String fileName);
+
+  /**
+   * Get an {@link OutputFile} for writing bytes to a new data file for this table.
+   * <p>
+   * The partition values of the rows in this file may be used to derive the final location of
+   * the file.
+   */
+  OutputFile newPartitionedDataFile(
```
Comment (mccheah) [inline api/src/main/java/com/netflix/iceberg/io/FileIO.java]: I was thinking about this a little bit more, and in contrast to what was discussed on the ticket, I propose creating the `OutputFile` instance directly instead of just returning the path alone.

This comes up in the case where the file doesn't exist yet, which is always going to be the case for Spark (Spark is always creating new data files). In such a case, it is desirable for the plugin to not necessarily only know what the path is, but also to perhaps create the path itself, oftentimes by asking for some new entry in an external system that is a reference to this file we're about to write to.

In other words, the semantics of this API are that we're not only asking for the location of the data file, but we're also asking to create the data file as part of this action. In practice most of the implementations should just derive this as `newOutputFile(getPath(args))`. Nevertheless I think this version of the API is more rich. Thoughts?

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java [hunk 2]
```diff
@@ -38,4 +55,98 @@ public void deleteFile(String path) {
...
+  @Override
+  public OutputFile newUnpartitionedDataFile(String fileName) {
+    return newOutputFile(String.format("%s/%s", newDataFileLocation, fileName));
+  }
+
+  private Path metadataPath(String filename) {
+    return new Path(new Path(tableLocation, "metadata"), filename);
+  }
+
+  public void updateProperties(Map<String, String> newTableProperties) {
```
Comment (mccheah) [inline core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java]: This is a concession to the fact that when table properties are updated, it's possible that the data path could change. It's not immediately clear if we can do better than this.

File: api/src/main/java/com/netflix/iceberg/io/FileIO.java (hunk 1 above)
Comment (rdblue) [reply api/src/main/java/com/netflix/iceberg/io/FileIO.java]: Sounds reasonable to me, as long as this doesn't cause odd practices like generating an OutputFile, converting it to a location, then later getting an OutputFile for it. I don't think that would happen though, so this is probably saving a call.

Comment (mccheah) [reply api/src/main/java/com/netflix/iceberg/io/FileIO.java]: We're often using the location of the `OutputFile` object, but we shouldn't be converting twice anywhere.

File: core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java [hunk 3]
```diff
@@ -52,8 +47,8 @@
   private static final String HIVE_LOCATION_FOLDER_NAME = "empty";
 
   private final Configuration conf;
-  private final FileIO fileIo;
 
+  private HadoopFileIO fileIo;
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: I don't think this needs to be a HadoopFileIO, does it?

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java (hunk 2 above)
Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java]: Actually `Supplier` isn't great because the `Supplier` itself isn't serializable. The `Supplier` would be tied to the `HadoopTableOperations` object.

File: api/src/main/java/com/netflix/iceberg/io/FileIO.java [hunk 4]
```diff
@@ -47,4 +46,30 @@
    * Delete the file at the given path.
    */
   void deleteFile(String path);
+
+  /**
+   * Get an {@link InputFile} to get the bytes for this table's metadata file with the given name.
+   */
+  InputFile readMetadataFile(String fileName);
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/io/FileIO.java]: Metastore tables always use full paths to metadata files, so this is only used by file system tables in `HadoopTables`. I think the need for this method is caused by not separating the concern of building a metadata path from getting an `InputFile` or `OutputFile` instance. It seems odd to me to add a method to this interface just for Hadoop tables. Another minor issue is that it isn't obvious why this takes a filename and not a full path: the name `readMetadataFile` doesn't indicate that to me. If a caller has a full path to a metadata file, why indicates that `readMetadataFile` is not the right choice?

On the other hand, if this were replaced by `metadataFileLocation`, it would still only be used by Hadoop tables because it is built into `newMetadataFile`. I like the way `newMetadataFile` and `newPartitionedDataFile` work, so I'd rather not change those.

What about removing `readMetadataFile` and instead using `newMetadataFile(filename).toInputFile()`? I think that it is clear that `newMetadataFile` adds the metadata path because it is distinct from `newOutputFile` and how to construct the full file path is a concern. I think this is a reasonable way to "guess" the location of metadata files for Hadoop tables.

File: api/src/main/java/com/netflix/iceberg/io/FileIO.java [hunk 5]
```diff
@@ -47,4 +46,30 @@
...
+
+  /**
+   * Get an {@link OutputFile} to write bytes for a new table metadata file with the given name.
+   */
+  OutputFile newMetadataFile(String fileName);
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/io/FileIO.java]: What do you think about renaming this to `newMetadataOutputFile` to be clear that this is producing an `OutputFile` and to fit the existing convention where `newInputFile` and `newOutputFile` are distinguished from one another?

File: api/src/main/java/com/netflix/iceberg/io/FileIO.java [hunk 6]
```diff
@@ -47,4 +46,30 @@
...
+
+  /**
+   * Get an {@link OutputFile} for writing bytes to a new data file for this table.
+   * <p>
+   * The partition values of the rows in this file may be used to derive the final location of
+   * the file.
+   */
+  OutputFile newPartitionedDataFile(
+      PartitionSpec partitionSpec, StructLike partitionData, String fileName);
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/io/FileIO.java]: It seems strange to me to pass in the partition spec and partition data. Why not have a single method, `newDataOutputFile`, that accepts a relative path?

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java (hunk 2 above)
Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java]: That makes sense. I think the best option is what you originally had, but with `TableMetadata`, `updateTableMetadata`. That's `Serializable` right?

Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java]: You could also make `tableMetadataSupplier` transient and cache the result. Then at least the FileIO pulls metadata when needed instead of requiring a push.

Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java]: Okay, so `HadoopFileIO` would need to extract paths from `TableMetadata` each time `updateMetadata` is called instead of storing a reference. I think it is still better to pass the whole `TableMetadata` in case anything else changes. Probably also note in javadoc that `TableMetadata` isn't serializable, so don't save a copy of it.

Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java]: `updateMetadata` is a fitting compromise here!

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java [hunk 7]
```diff
@@ -69,30 +73,33 @@ public TableMetadata current() {
...
     int ver = version != null ? version : readVersionHint();
-    Path metadataFile = metadataFile(ver);
-    FileSystem fs = getFS(metadataFile, conf);
+    InputFile metadataFile = io().readMetadataFile(metadataFileName(ver));
+    Path metadataFilePath = new Path(metadataFile.location());
+    FileSystem fs = getFS(metadataFilePath, conf);
     try {
       // don't check if the file exists if version is non-null because it was already checked
-      if (version == null && !fs.exists(metadataFile)) {
+      if (version == null && !fs.exists(metadataFilePath)) {
```
[the remark this answers is not in the document]
Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java]: Yeah I also understand not wanting to add `exists`. I'm also hesitant because we don't want to blow up these APIs to become a full-fledged file system. Where is the boundary with regards to what File System-like operations we will require implementations to provide?

File: api/src/main/java/com/netflix/iceberg/io/FileIO.java (hunk 6 above)
Comment (mccheah) [reply api/src/main/java/com/netflix/iceberg/io/FileIO.java]: Passing in specifically the partitioning information here I think carries a stronger notion that this will be used for data with those partitionings, as opposed to a path in which callers can really pass in any string. So the stronger typing carries stronger semantics. It also signifies that the partitioning need not be encoded in the path, but the implementation can choose to derive its own paths based on the partitioning.

Comment (rdblue) [reply api/src/main/java/com/netflix/iceberg/io/FileIO.java]: > partitioning need not be encoded in the path

Good point. What about naming both `newDataOutputFile`? The arguments still carry the notion that this is for the partitioned case.

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java]: We want to keep it as light-weight as possible to make it easy to avoid using the Hadoop API in the future. There's no reason to force downstream apps to include it. After that goal is met, then it's really a matter of trying to do as much as we can cleanly through the API. I think that adding exists would be okay. Then the only non-standard part of the Hadoop implementation would be the rename call.

File: api/src/main/java/com/netflix/iceberg/io/FileIO.java (hunk 6 above)
Comment (mccheah) [reply api/src/main/java/com/netflix/iceberg/io/FileIO.java]: Do we need the second unpartitioned variant? For example, what if we passed `PartitionSpec#unpartitioned` and an empty `PartitionData` structure instead? Or do we think that the API should make unpartitioned files a first-class concept?

**Comment by mccheah** [comment]
@rdblue addressed most of the comments. Think there's a few things we can continue to think about. Also worth highlighting is that the last commit is a conflict resolution where I also moved the metadata location configuration into `HadoopFileIO` but it's not clear if I did that correctly, though tests pass.

File: api/src/main/java/com/netflix/iceberg/io/FileIO.java (hunk 6 above)
Comment (rdblue) [reply api/src/main/java/com/netflix/iceberg/io/FileIO.java]: I think it is easier for implementers to have two methods. We could add a default implementation that calls the partitioned version and let implementations choose, but I think it is much easier to implement the unpartitioned case without worrying about the spec.

File: api/src/main/java/com/netflix/iceberg/io/FileIO.java [hunk 8]
```diff
@@ -47,4 +46,25 @@
    * Delete the file at the given path.
    */
   void deleteFile(String path);
+
+  /**
+   * Get an {@link OutputFile} to write bytes for a new table metadata file with the given name.
+   */
+  OutputFile newMetadataOutputFile(String fileName);
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/io/FileIO.java]: Please include `@param filename` that explicitly points out that this should not include a path.

File: api/src/main/java/com/netflix/iceberg/io/FileIO.java [hunk 9]
```diff
@@ -47,4 +46,25 @@
...
+
+  /**
+   * Get an {@link OutputFile} for writing bytes to a new data file for this table.
+   * <p>
+   * The partition values of the rows in this file may be used to derive the final location of
+   * the file.
+   */
+  OutputFile newDataOutputFile(
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/io/FileIO.java]: Same here, I think `@param` docs will help clarity. Also adding an example to get the path from the spec and partition data would help.

File: core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java [hunk 10]
```diff
@@ -84,13 +75,24 @@ protected void requestRefresh() {
   }
 
   protected String writeNewMetadata(TableMetadata metadata, int version) {
-    String newTableMetadataFilePath = newTableMetadataFilePath(metadata, version);
-    OutputFile newMetadataLocation = fileIo.newOutputFile(newTableMetadataFilePath);
+    FileIO initializedFileIo = io();
+    // This would be false if the subclass overrides io() and returns a different kind.
+    if (initializedFileIo instanceof HadoopFileIO) {
+      ((HadoopFileIO) initializedFileIo).updateTableMetadata(metadata);
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: Alternatively, we could pass TableMetadata to `newMetadataOutputFile` and `newDataOutputFile`. Then we wouldn't need this notification.

Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: The problem is that `TableMetadata` is in `iceberg-core` but this module is in `iceberg-api`. But `FileIO` is returned by `Table`, which is in `iceberg-api`. Thus to make this API take `TableMetadata` as input we'd likely have to move `TableMetadata` into `iceberg-api`. The modules aren't set up correctly right now.

Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: Moving `TableMetadata` into `iceberg-api` runs into the problem of `TableOperations` not being in iceberg-api.

Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: I think better would be that instead of caching the `FileIO` object and trying to maintain its state, we should just make a new one with the most up to date metadata every time it's needed to write something.

Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: > I think we would expose the classes themselves. We'd add an interface to the API and implement that. That gets around the TableOperations problem.

Meaning, add an interface on top of `TableMetadata`?

**Comment by mccheah** [comment]
Addressed comments.

File: core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java [hunk 11]
```diff
@@ -114,30 +103,19 @@ protected void refreshFromMetadataLocation(String newLocation, int numRetries) {
...
-  public String metadataFileLocation(String filename) {
-    return metadataFileLocation(current(), filename);
+  public FileIO io() {
+    return new HadoopFileIO(conf, currentMetadata);
   }
 
   @Override
-  public FileIO io() {
-    return fileIo;
+  public FileIO io(TableMetadata tableMetadata) {
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: Is it possible to make this private? It is odd for a TableOperations class that provides TableMetadata instances to have a public method that accepts TableMetadata instead of using current.

Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: Sorry, I'm now looking at the javadoc for it.

Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: Okay, looks like the trade-off here is that BaseTransaction needs to provide a FileIO object and its view of TableMetadata may be different from the table's actual metadata state because we stack up multiple changes and then commit just the final version in the transaction.

So unless I'm missing something again, the only thing that adding this method accomplishes is to be able to update some location and have the change take effect for a write in the same atomic commit. I'm okay with that limitation to make this API cleaner, unless you need to be able to change settings and write in an atomic commit for your use case.

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java [hunk 12]
```diff
@@ -1,21 +1,106 @@
...
 
-  public HadoopFileIO(Configuration hadoopConf) {
+  public HadoopFileIO(
+      Configuration hadoopConf, TableMetadata tableMetadata) {
     this.hadoopConf = new SerializableConfiguration(hadoopConf);
+    this.tableLocation = tableMetadata.location();
+    Map<String, String> tableProperties = tableMetadata.properties();
+    this.newDataFileDir = stripTrailingSlash(
+        tableProperties
+            .getOrDefault(
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/hadoop/HadoopFileIO.java]: Nit: different indentation than the next statement? (Not a big deal if you aren't making further changes)

File: core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java [hunk 13]
```diff
@@ -109,32 +104,30 @@ public void commit(TableMetadata base, TableMetadata metadata) {
       return;
     }
 
+    FileIO newMetadataIO = io(metadata);
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/hadoop/HadoopTableOperations.java]: Why move this before the validation checks? If those fail, this file would need to be cleaned up.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 14]
```diff
@@ -327,11 +261,11 @@ public WriterCommitMessage commit() throws IOException {
       close();
 
       if (metrics.recordCount() == 0L) {
-        fileIo.deleteFile(file);
+        fileIo.deleteFile(file.location());
         return new TaskCommit();
       }
 
-      InputFile inFile = fileIo.newInputFile(file);
+      InputFile inFile = fileIo.newInputFile(file.location());
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: What about `file.toInputFile()`?

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 15]
```diff
@@ -423,7 +356,7 @@ public void abort() throws IOException {
       if (currentAppender != null) {
         currentAppender.close();
         this.currentAppender = null;
-        fileIo.deleteFile(currentPath);
+        fileIo.deleteFile(currentFile.location());
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: Maybe it would make sense to add `deleteFile(InputFile file)` with a default implementation that calls `file.location()`.

**Comment by rdblue** [comment]
@mccheah, thanks for working on this. It's looking close, but I have doubts about adding `TableOperations#io(TableMetadata)`. I'd rather add `FileIO#updateMetadata(TableMetadata)` if you think it is necessary to have the behavior that requires `io(TableMetadata)`.

Otherwise, I think this is about ready to get in.

File: core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java (hunk 11 above)
Comment (mccheah) [reply core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java]: > So unless I'm missing something again, the only thing that adding this method accomplishes is to be able to update some location and have the change take effect for a write in the same atomic commit. I'm okay with that limitation to make this API cleaner, unless you need to be able to change settings and write in an atomic commit for your use case.

So I accidentally stumbled upon having to do it this way because a lot of tests just failed if we used the old location to commit when the new metadata specifies committing elsewhere. Thus I had thought that the current behavior is part of public API and we don't intend on changing it. I suppose we ought to document this specific behavior somewhere (if you change the new metadata location table property on a commit, the change isn't reflected in that current commit). However it's unclear where the right place to document that is.

**Comment by mccheah** [comment]
To override FileIO generation for subclasses of `BaseMetastoreTableOperations` though, devs should override `BaseMetastoreTableOperations#metadataAwareFileIO`.

**Comment by mccheah** [comment]
Superceded by https://github.com/apache/incubator-iceberg/pull/87
